fp8 backward

micmelesse opened the pull request:
add fp8 backward

brunomazzottiamd commented on flash_attn/flash_attn_triton_amd/bwd_prefill.py:
```diff
@@ -223,48 +243,71 @@ def _bwd_kernel_one_col_block(
+        ds = dscores_scaled * sm_scale
+        ds = tl.where(p_mask, ds, 0.0)
+        
+        # print("p:", p)
```
Can we clean up these `print` statements before merging?

brunomazzottiamd commented on flash_attn/flash_attn_triton_amd/bwd_prefill.py:
```diff
@@ -223,48 +243,71 @@ def _bwd_kernel_one_col_block(
+        # print("delta_i:", delta_i)
+        # print("ds:", ds) # NOTE:is almost the same between fp8 and fp16
+        descale_ds = descale_p
+        ds_fp8_scaled = ds * (1.0/ descale_ds)
```
We only need to compute `ds_fp8_scaled` for fp8 kernel, otherwise it's unused. What do you think of computing `ds_fp8_scaled` inside a `if IS_FP8:` statement?

brunomazzottiamd commented on flash_attn/flash_attn_triton_amd/bwd_prefill.py:
```diff
@@ -553,6 +636,14 @@ def attention_prefill_backward_triton_impl(
         print("sequence_parallel:", sequence_parallel)
 
+    is_fp8 = arch_supports_fp8() and q.dtype in {torch.float8_e4m3fnuz, torch.float8_e4m3fn, torch.float8_e5m2, torch.float8_e5m2fnuz}
+    if is_fp8:
```
I think this empty `if is_fp8:` statement can be removed. Do we need to print or debug anything inside it?

brunomazzottiamd replied:
I'm marking this discussion as **unresolved** because I think the empty `if` statement is still here.

micmelesse replied:
we are not working bwd_prefill now.

brunomazzottiamd commented on flash_attn/flash_attn_triton_amd/test.py:
```diff
@@ -785,16 +786,21 @@ def test_op_prefill_fp8(Z, HQ, HK, N_CTX_Q, N_CTX_K, D_HEAD, causal, dropout_p,
-            k_bfp16,
-            v_bfp16,
+    # ref forward pass
+    # NOTE: bfp16 is not supported by atomic ops. we have to use fp16
```
I was about to ask what did you do with bf16 atomics...

brunomazzottiamd commented on flash_attn/flash_attn_triton_amd/test.py:
```diff
@@ -845,17 +853,55 @@ def test_op_prefill_fp8(Z, HQ, HK, N_CTX_Q, N_CTX_K, D_HEAD, causal, dropout_p,
+    if DEBUG:
+        print("S_dmask_fp16:", S_dmask_fp16, S_dmask_fp16.shape if S_dmask_fp16 is not None else None )
+        print("S_dmask_fp8:", S_dmask_fp8, S_dmask_fp8.shape if S_dmask_fp16 is not None else None)
+    torch.testing.assert_close(S_dmask_fp16.to(torch.float32) if S_dmask_fp16 is not None else None, S_dmask_fp8.to(torch.float32) if S_dmask_fp8 is not None else None, atol=ATOL_fp8, rtol=RTOL_fp8)
```
Does `torch.testing.assert_close` handle `None` arguments well?

micmelesse replied:
yes, when I run the tests it is fine

brunomazzottiamd commented on flash_attn/flash_attn_triton_amd/test.py:
```diff
@@ -845,17 +853,55 @@ def test_op_prefill_fp8(Z, HQ, HK, N_CTX_Q, N_CTX_K, D_HEAD, causal, dropout_p,
+    torch.testing.assert_close(S_dmask_fp16.to(torch.float32) if S_dmask_fp16 is not None else None, S_dmask_fp8.to(torch.float32) if S_dmask_fp8 is not None else None, atol=ATOL_fp8, rtol=RTOL_fp8)
+
+    if HQ // HK != 1:
+        print("Skipping backward for MQA/GQA cases because atomic_add doesnot support fp8")
```
I was about to ask what did you do fp8 MQA/GQA...

micmelesse replied:
yeah, @jtang10 kernel doesnot use atomics as far as I am aware so the point of this pr is just to figure out how to do fp8 backward. People will be using his kernel finally.

brunomazzottiamd replied:
We're 100% in the same page.

brunomazzottiamd reviewed (approved):
I'm approving the PR because I can't see anything wrong with it. I just left some questions and cleanup suggestions.

micmelesse commented on flash_attn/flash_attn_triton_amd/README.md:
````diff
@@ -43,6 +43,9 @@ python setup.py install
 ```
 
+##### FP8
+In our fork, we have modified the api to work with fp8. You provide tensors that are scaled to be in fp8 range and their associated descaling factors.
````
I am going to add more info to the README. This was just the start.

brunomazzottiamd commented on flash_attn/flash_attn_triton_amd/fwd_prefill.py:
```diff
@@ -693,15 +687,15 @@ def attention_prefill_forward_triton_impl(
                     BLOCK_DMODEL=padded_d_model, USE_BIAS=False if bias is None else True,
                     USE_ALIBI=False if alibi_slopes is None else True, ENABLE_DROPOUT=dropout_p
-                    > 0.0, USE_EXP2=use_exp2, RETURN_SCORES=return_softmax, IS_FP8=is_fp8)
+                    > 0.0, USE_EXP2=use_exp2, RETURN_SCORES=return_softmax, IS_FP8=is_fp8, FP8_MAX=torch.finfo(torch.float8_e4m3fnuz).max)
```
**Suggestion:**

Since `is_fp8` is defined as:
```python
is_fp8 = arch_supports_fp8() and q.dtype in {torch.float8_e4m3fnuz, torch.float8_e4m3fn, torch.float8_e5m2, torch.float8_e5m2fnuz}
```
I think it's better to compute `FP8_MAX` as:
```python
FP8_MAX=torch.finfo(q.dtype).max
```

brunomazzottiamd commented on flash_attn/flash_attn_triton_amd/bwd_prefill.py:
```diff
@@ -650,56 +741,26 @@ def attention_prefill_backward_triton_impl(
         delta,
         stride_oz, stride_oh, stride_om, stride_ok,
-        stride_oz, stride_oh, stride_om, stride_ok,
+        stride_oz, stride_oh, stride_om, stride_ok, # FIXME: don't share strides with derivatives this was causing a lot of issues
```
**Question:** Should we merge this PR without addressing this `FIXME:`? Is it supposed to be addressed later?

brunomazzottiamd commented on flash_attn/flash_attn_triton_amd/bwd_prefill.py:
```diff
@@ -741,19 +806,23 @@ def attention_prefill_backward_triton_impl(
+        IS_VARLEN=is_varlen,
+        GROUP_SIZE=group_size,
+        IS_FP8=is_fp8,
+        FP8_MAX=torch.finfo(torch.float8_e4m3fnuz).max
```
**Suggestion:**

Since `is_fp8` is computed as:
```python
is_fp8 = arch_supports_fp8() and q.dtype in {torch.float8_e4m3fnuz, torch.float8_e4m3fn, torch.float8_e5m2, torch.float8_e5m2fnuz}
```

I think it's better to compute `FP8_MAX` as:`
```python
FP8_MAX=torch.finfo(q.dtype).max
```

brunomazzottiamd commented on flash_attn/flash_attn_triton_amd/bwd_prefill_split.py:
```diff
@@ -897,18 +1006,31 @@ def attention_prefill_backward_triton_split_impl(
 ):
+    IS_FP8 = arch_supports_fp8() and q.dtype in {torch.float8_e4m3fnuz, torch.float8_e4m3fn, torch.float8_e5m2, torch.float8_e5m2fnuz}
+    if IS_FP8:
+        FP8_MAX = torch.finfo(torch.float8_e4m3fnuz).max
```
**Suggestion:** Compute `FP8_MAX` as `torch.finfo(q.dtype).max`.

brunomazzottiamd commented on flash_attn/flash_attn_triton_amd/test.py:
```diff
@@ -849,57 +856,90 @@ def test_op_prefill_fp8(Z, HQ, HK, N_CTX_Q, N_CTX_K, D_HEAD, causal, dropout_p,
-        (4, 8, 8, 2048, 2048, 128),
-        (4, 16, 16, 4096, 4096, 64),
-        (2, 4, 4, 8192, 8192, 32),
+        # (1, 1, 1, 1, 1, 1),
```
**Question:** Is it our intention to leave only one test case? Or is this still a work in progress?

brunomazzottiamd commented on .github/workflows/amd_tests.yml:
```diff
@@ -50,33 +50,33 @@ jobs:
-        run: |
-          export FLASH_ATTENTION_TRITON_AMD_ENABLE="TRUE"
-          pytest tests/test_flash_attn_triton_amd.py
+      # - name: Flash Attention Tests Using Reference Impl
```
**Question:** My knowledge of GitHub actions is almost none, so please take this comment with a grain of salt... As far as I can see, MI300 integration job is commented out. Am I correct? Do we really want to merge this way?

brunomazzottiamd reviewed (approved):
Great job Michael! Kudos for introducing `compute_fp8_scaling_factors` Triton function, it's really useful to avoid code repetition.

micmelesse committed:
fp8 BWD after figuring out varlen problem

This is a combination of 21 commits.

fp8 BWD

Enable BWD fp8 with split kernel

Enable BWD fp8 with per block scale factors for p
and ds

This is a combination of 9 commits.

Enable BWD fp8

This is a combination of 12 commits.

add backward test case

save clean up

disable ci

lse is good

dv matches

reduce diff

use do fp8 for dv

kinda working

group size is a constexpr

clean up a bit

everything except mqa/gqa works

skip mqa cases

20 cases have nan on dropout

save what you have

disable tests

failing

enable tests

per block descale_p and descale_ds

use max(abs(())

clean up tests a bit more

fix bug

disable ci for now

pass variables

add flags

add alternate path. Still need to load descale factors

dv working

dk works

save

add type info for backward

fix  DEBUG flag bug

fix bug with backward. Normal forward works with dropout. Segfault with causal. Varlen has some issues. Might be related to strides.

pass descale strides

test causal

fix causal compiler assert. min head should be 32

remove descale_p

save

explict name as causal

isolate bad case

just run fp8 tests

bench with autotune

min changes

cast_fp8 helper

cast_varlen_to_fp8

save

minor

highlight failing configs

increase test cases

mark failing

recategorize misc tests

group failing gqa configs

add more tests

add vis code

min ci changes

dump folder

single image per tensors

add tensor comparison

gen varlen tensor

vis varlen tensors

varlen diff

nice varlen vis

vis function

show seqlen in varlen

add vis_tensors function

simplify

add color bars

rm vis from test

set canvas size.

descale values are optional

add ck tests

add flag to build ck

rm ck test

assert requires grad

ensure q, k, and v require gradients

split vis

rm interp, 8k and 300 dpi

slice per page

disable ci for now

add more vis code

tensor per image is better

for vis_close, don't vis if no error. also vis all failing varlen tests

varlen failures due to different seqlens

rm vis code

micmelesse committed:
rm require grad

micmelesse committed:
decast fp8 for ref input, use fp16 as input

fix minor things

match readme

decast fp8 for ref input, use fp16 as input

micmelesse committed:
disable causal

micmelesse committed:
fix bug

micmelesse committed:
pass strides

micmelesse committed:
DEBUG modes work only with interp

micmelesse committed:
zero out varlen bwd grads

micmelesse committed:
zero out everything

micmelesse committed:
varlen dropout and causal works

micmelesse committed:
add descale factors to other apis

micmelesse committed:
save

micmelesse committed:
unify tests

micmelesse committed:
add packing flag

micmelesse committed:
fix copy grad bug

micmelesse committed:
add types, flags for zeroing tensors and
accumlating fp32

This is a combination of 5 commits.

extend ci time

clean more

minimize difference

add types

ZERO_TENSORS and ACCUMLATE_FP32 flags

micmelesse committed:
just pass the output tensors

micmelesse committed:
accumlate forwad in fp32

micmelesse committed:
fp8 in and fp8 out

micmelesse committed:
return descale factors works for out

micmelesse committed:
start fp8 return for bwd

micmelesse committed:
return dq, dv, dk descale factors

micmelesse committed:
save what you have

micmelesse committed:
custom fp8 api function

micmelesse committed:
add varlen function

micmelesse committed:
test backward with varlen

micmelesse committed:
test fp8

micmelesse committed:
kv cache fix

micmelesse committed:
clean up interface

micmelesse committed:
add packed api

micmelesse committed:
fix qkv bug

micmelesse commented:
CI has been flaky all night but it is running now. This are my local results

main tests
```
======== 29094 passed, 12678 skipped, 1 warning in 12202.62s (3:23:22) =========
```

fp8 tests
=========== 616 passed, 280 skipped, 1 warning in 454.98s (0:07:34) ============

micmelesse committed:
disable bench

micmelesse committed:
run big tests at the end

micmelesse committed:
run in parrallel

micmelesse committed:
Update utils.py

micmelesse committed:
Update amd_tests.yml

micmelesse committed:
add train script

micmelesse commented:
I trained a very small lm on wikitext using both the normal api and the fp8 api. Things seem good to me. It is overfit but for my purpose of showing no numeric issues. I think this is good enough. I have add the script that is used to train the model as train.py. Ci has been flaky so I am going to run a local check on the final commit and merge the pr.

![validation_loss](https://github.com/user-attachments/assets/457be4c6-c1b8-4021-8d7f-7f66b7834fcd)
![training_loss](https://github.com/user-attachments/assets/61930291-dd57-4e72-aeb7-1e223ad52dad)

micmelesse committed:
use local configs for testing
